### ISSUE TITLE
logger.catch exception

### ISSUE THREAD

**Issue by m3lab-zzl** [Delgan/loguru#1108]
https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.catch

I love this wrapper, but it gets me confused when handling errors. 

```python
from loguru import logger
import sys
import os


@logger.catch(onerror=lambda _: sys.exit(1))
def get_existing_file_path(fileName: str) -> str:
    """Check if fileName points to an existing file and return its absolute path"""
    assert isinstance(
        fileName, str
    ), f"fileName must be a string, but it's of type {type(fileName)}"
    absf = os.path.abspath(fileName)
    if os.path.isfile(absf):
        return absf
    else:
        raise FileNotFoundError(f"{absf} file not found")


try:
    get_existing_file_path("nothing")
except FileNotFoundError:
    logger.warning("Generating xyz file...")
```

catch() returns wrapper so that the traditional `FileNotFoundError` is wiped? I want to show variables in `get_existing_file_path` ONLY when there are errors not handled by the outside code. Is it possible?

**Comment by changchiyou** [comment]
```diff
from loguru import logger
import sys
import os

- @logger.catch(onerror=lambda _: sys.exit(1))
+ @logger.catch(reraise=True)
def get_existing_file_path(fileName: str) -> str:
    """Check if fileName points to an existing file and return its absolute path"""
    assert isinstance(
        fileName, str
    ), f"fileName must be a string, but it's of type {type(fileName)}"
    absf = os.path.abspath(fileName)
    if os.path.isfile(absf):
        return absf
    else:
        raise FileNotFoundError(f"{absf} file not found")


try:
    get_existing_file_path("nothing")
except FileNotFoundError:
    logger.warning("Generating xyz file...")
```

<img width="885" alt="image" src="https://github.com/Delgan/loguru/assets/46549482/0f889643-7070-4fd2-850f-5c9f037e0260">

**Comment by m3lab-zzl** [comment]
@changchiyou Many thanks, I will try it soon

**Comment by m3lab-zzl** [comment]
@changchiyou , with `reraise`, the error can be caught by outside code, but it also shows the needless FileNotFound error message. 

I want something that is as similar as possible to standard python output (without loguru), which only shows the traced variable when there is something unexpected happened.

**Comment by changchiyou** [comment]
> @changchiyou , with `reraise`, the error can be caught by outside code, but it also shows the needless FileNotFound error message. 
> 
> I want something that is as similar as possible to standard python output (without loguru), which only shows the traced variable when there is something unexpected happened.

I can't understand what you want: could you please provide some expected results?

**Comment by Delgan** [comment]
Maybe you don't need `@logger.catch()` at all?

```python
from loguru import logger
import sys
import os


def get_existing_file_path(fileName: str) -> str:
    """Check if fileName points to an existing file and return its absolute path"""
    assert isinstance(
        fileName, str
    ), f"fileName must be a string, but it's of type {type(fileName)}"
    absf = os.path.abspath(fileName)
    if os.path.isfile(absf):
        return absf
    else:
        raise FileNotFoundError(f"{absf} file not found")


try:
    get_existing_file_path("nothing")
except FileNotFoundError:
    logger.opt(exception=True).warning("Generating xyz file...")
```

**Comment by m3lab-zzl** [comment]
Sorry, let me explain this.

Outside code only handles FileNotFoundError, while the internal function may also raise other exceptions, such as AssertionError:

```python
from loguru import logger
import sys
import os


@logger.catch()
def get_existing_file_path(fileName: str) -> str:
    """Check if fileName points to an existing file and return its absolute path"""
    assert isinstance(fileName, str)
    absf = os.path.abspath(fileName)
    if os.path.isfile(absf):
        return absf
    else:
        raise FileNotFoundError(f"{absf} file not found")


try:
    get_existing_file_path(1)
except FileNotFoundError:
    logger.opt(exception=True).warning("Generating xyz file...")
```

![image](https://github.com/Delgan/loguru/assets/102862066/f45d958f-40f3-42d9-bd5c-dc6f049d56c3)

`@logger.catch()` decorator reports the value of `fileName`, which is super helpful. Without it, I can only see this:

![image](https://github.com/Delgan/loguru/assets/102862066/ac28f66a-f9e4-4556-a85c-e9f6c3f61c4a)

Thus, maybe I should keep `@logger.catch()`. However, I do not need the FileNotFoundError message since it is handled by the outside try block:

```python
try:
    get_existing_file_path("something")
except FileNotFoundError:
    logger.opt(exception=True).warning("Generating xyz file...")
```

In short, no `catch`, no variable diagnose:
![image](https://github.com/Delgan/loguru/assets/102862066/69998806-a0d3-447d-b786-bd0f5f160b00)

with `catch`, can't get rid of "handled" exception:
![image](https://github.com/Delgan/loguru/assets/102862066/54023f40-d10c-4f65-a1ab-44152d3875e0)

![image](https://github.com/Delgan/loguru/assets/102862066/bf69765d-0546-49c2-a775-e4470a7cc644)

I guess my code needs some sort of refactoring to use `catch` properly. Could you please give me some advice?

**Comment by Delgan** [comment]
I see. Maybe use `@logger.catch(exclude=FileNotFoundError)`?

**Comment by m3lab-zzl** [comment]
Thanks for the hint. It works as expected:

```python
from loguru import logger
import sys
import os

@logger.catch(exclude=FileNotFoundError)
def get_existing_file_path(fileName: str) -> str:
    """Check if fileName points to an existing file and return its absolute path"""
    assert isinstance(fileName, str)
    absf = os.path.abspath(fileName)
    if os.path.isfile(absf):
        return absf
    else:
        raise FileNotFoundError(f"{absf} file not found")


try:
    get_existing_file_path("hi")
except FileNotFoundError:
    logger.warning("Generating xyz file...")
```

![image](https://github.com/Delgan/loguru/assets/102862066/31c5f071-7af3-4c40-a927-a96346b16b2e)
